File: Python/ASP Advanced/Linked Lists/Fast and Slow Pointers/Leetcode-141.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:


        if not head:
            return False
        
        
        slow, fast = head, head.next

        while fast and fast != slow and fast.next:
            slow = slow.next
            fast = fast.next.next

        return True if fast == slow else False

File: ASP Advanced/Linked Lists/Leetcode-141.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode* fast = head->next, * slow = head;

        while (fast && fast != slow){
            fast = fast->next;
            if (fast) fast = fast->next;
            slow = slow->next;
        }
        return (!fast)? false : true;
    }
};
// The code defines a solution to the problem of detecting a cycle in a linked list.
// It uses a two-pointer technique (Floyd's Cycle Detection Algorithm) to determine if a cycle exists in the list.
// The function returns true if a cycle is detected, otherwise it returns false.